modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/partitionawareness/PartitionAwarenessMetadataExtractor.java

```diff
@@ -113,11 +124,31 @@ public static PartitionAwarenessMetadata getMetadata(IgniteKeyValueModify kv) {
             if (expr instanceof RexDynamicParam) {
                 RexDynamicParam dynamicParam = (RexDynamicParam) expr;
                 indexes[i] = dynamicParam.getIndex();
+                onlyLiterals = false;
+            } else if (expr instanceof RexLiteral) {
+                RexLiteral expr0 = (RexLiteral) expr;
+
+                // depends on supplied zoneId, it can`t be cached
+                if (expr0.getTypeName() == SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE) {
+                    return null;
+                }
+
+                indexes[i] = hashPos--;
+                NativeType nativeType = IgniteTypeFactory.relDataTypeToNative(expr0.getType());
+                Object val = getValueFromLiteral(nativeType, expr0);
+                hashFields.add(ColocationUtils.hash(val, nativeType));
             } else {
                 return null;
             }
         }
 
+        // case for IgniteKeyValueModify with only literals in colocation columns.
+        if (fullRow && onlyLiterals) {
+            return null;
```

Comment:
what is the point of discarding expensive computation here?

Reply:
done